client/state/http/index.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import superagent from 'superagent';
+
+/***
+ * Internal dependencies
+ */
+import { extendAction } from 'state/utils';
+import { HTTP_REQUEST } from 'state/action-types';
+import { failureMeta, successMeta } from 'state/data-layer/wpcom-http';
+
+const isAllHeadersValid = headers =>
+	headers.every( headerPair => Array.isArray( headerPair ) &&
+		headerPair.length === 2 &&
+		typeof headerPair[ 0 ] === 'string' &&
+		typeof headerPair[ 1 ] === 'string'
+	);
+
+/***
+ * Handler to perform an http request based on `HTTP_REQUEST` action parameters:
+ * {String} url the url to request
+ * {String} method the method we should use in the request: GET, POST etc.
+ * {Array<Array<String>>} headers array of [ 'key', 'value' ] pairs for the request headers
+ * {Object} body data send as body
+ * {Boolean} withCredentials save cookie set on request
+ * {Action} onSuccess action to dispatch on success with data meta
+ * {Action} onFailure action to dispatch on failure with error meta
+ *
+ * @param {Function} dispatch redux store dispatch
+ * @param {Object} action dispatched action we need to handle
+ * @returns {Promise} promise of the handled request
+ */
+const httpHandler = ( { dispatch }, action ) => {
+	const {
+		url,
+		method,
+		headers = [],
+		body,
+		withCredentials,
+		onSuccess,
+		onFailure
+	} = action;
+
+	if ( ! isAllHeadersValid( headers ) ) {
+		const error = new Error( "Not all headers were of an array pair: [ 'key', 'value' ]" );
+		dispatch( extendAction( onFailure, failureMeta( error ) ) );
+		return Promise.reject( error );
+	}
+
+	const request = superagent( method, url );
+
+	if ( withCredentials ) {
+		request.withCredentials();
+	}
+
+	headers.forEach( ( [ headerKey, headerValue ] ) => request.set( headerKey, headerValue ) );
+
+	request.accept( 'application/json' );
```

Comment:
are we then limiting this to JSON calls? what about images or text documents, or an XML api? could we move this maybe into the request object and provide JSON as the default?